src/connect.c

```diff
@@ -96,30 +98,69 @@ int W32_CALL connect (int s, const struct sockaddr *servaddr, socklen_t addrlen)
 
   if (socket->remote_addr)
   {
-    if ((socket->so_type == SOCK_STREAM) &&
-        (socket->so_state & SS_NBIO))
+    if (socket->so_type == SOCK_STREAM)
     {
-      tcp_tick (NULL);
-
-      if (_sock_pending_connect (socket))
+      if (socket->so_state & SS_ISCONNECTING)
       {
-        SOCK_DEBUGF ((", EALREADY"));
-        SOCK_ERRNO (EALREADY);
-        return (-1);
+        tcp_tick (NULL);
+
+        if (_sock_pending_connect (socket))
+        {
+          SOCK_DEBUGF ((", EALREADY"));
+          SOCK_ERRNO (EALREADY);
+          return (-1);
+        }
+
+        if (socket->so_error != 0)
+        {
+          SOCK_DEBUGF ((", SO_ERROR: %s", short_strerror(socket->so_error)));
+          SOCK_ERRNO (socket->so_error);
+          socket->so_error = 0;
+          return (-1);
+        }
       }
 
-      if (socket->so_error != 0)
-      {
-        SOCK_DEBUGF ((", SO_ERROR: %s", short_strerror(socket->so_error)));
-        SOCK_ERRNO (socket->so_error);
-        socket->so_error = 0;
-        return (-1);
-      }
+      SOCK_DEBUGF ((", connect already done!"));
+      SOCK_ERRNO (EISCONN);
+      return (-1);
     }
 
-    SOCK_DEBUGF ((", connect already done!"));
-    SOCK_ERRNO (EISCONN);
-    return (-1);
+    SOCK_DEBUGF ((", reconnecting"));
+
+    /* No need to reconnect if same peer address/port.
+     */
+    if (!is_ip6)
+    {
+      const struct sockaddr_in *ra = (const struct sockaddr_in*)socket->remote_addr;
+
+      if (addr->sin_addr.s_addr == ra->sin_addr.s_addr &&
+          addr->sin_port        == ra->sin_port)
+        return (0);
+    }
+#if defined(USE_IPV6)
+    else
+    {
+      const struct sockaddr_in6 *ra    = (const struct sockaddr_in6*)socket->remote_addr;
+
+      if (!memcmp(&addr6->sin6_addr, &ra->sin6_addr, sizeof(addr6->sin6_addr)) &&
+          addr6->sin6_port == ra->sin6_port)
+        return (0);
+    }
+#endif
+
+    free (socket->remote_addr);
+    socket->remote_addr = NULL;
```

Comment:
Would also be nice if this realloc can be avoided.  But I guess the `sockaddr` could be any size.

Reply:
> But I guess the sockaddr could be any size.

Pretty much yes. For `AF_UNIX`, the maximum would be `sizeof(struct sockaddr_un)` which is 110 by default 
(if `UNIX_PATH_MAX` is default). But a user could define this to anything I guess.

But since `AF_UNIX` sockets are unsupported in Watt-32, we could define the max as 
`sizeof(struct sockaddr_in6) == 24`.

Reply:
Good point, so we can just always `malloc(24)` basically.  Could also store it in a union maybe (but that would involve touching a *lot* of code).

Reply:
Is it even possible to have different address families in `local_addr` and `remote_addr` though?  That doesn't seem logical.

Reply:
I believe that true. Brain-fart on my part.

Reply:
No worries, I was confused about it too :)